drivers/hid/hidclass/fdo.c

```diff
@@ -388,6 +390,215 @@ HidClassFDO_GetDescriptors(
+    {
+        listEntry = RemoveHeadList(&deviceContext->PendingIRPList);
+        nextIrp = CONTAINING_RECORD(listEntry, IRP, Tail.Overlay.ListEntry);
+
+        oldCancelRoutine = IoSetCancelRoutine(nextIrp, NULL);
+        if (oldCancelRoutine != NULL)
+        {
+            // Cancel routine not called for this IRP. Return this IRP.
+            ASSERT(oldCancelRoutine == HidClassPDO_IrpCancelRoutine);
+        }
+        else
+        {
+            // IRP is canceled
+            ASSERT(nextIrp->Cancel);
+            InitializeListHead(&nextIrp->Tail.Overlay.ListEntry);
+            nextIrp = NULL;
+        }
+    }
+
+    return nextIrp;
+}
+
+NTSTATUS
+NTAPI
+HidClassFDO_ReadCompletion(
+    IN PDEVICE_OBJECT DeviceObject,
+    IN PIRP Irp,
+    IN PVOID Context)
+{
+    PHIDCLASS_FDO_EXTENSION DeviceExtension;
+    PHIDCLASS_PDO_DEVICE_EXTENSION PDODeviceExtension = NULL;
+    ULONG CollectionNumber;
+    KIRQL CurrentIRQL;
+    LIST_ENTRY completeIRP;
+
+    UNREFERENCED_PARAMETER(DeviceObject);
+
+    /* get device extension */
+    DeviceExtension = Context;
+
+    if (Irp->IoStatus.Status == STATUS_PRIVILEGE_NOT_HELD ||
+        Irp->IoStatus.Status == STATUS_DEVICE_NOT_CONNECTED ||
+        Irp->IoStatus.Status == STATUS_CANCELLED)
+    {
+        /* failed to read or should be stopped*/
+        DPRINT1("[HIDCLASS] ReadCompletion terminating read Status %x\n", Irp->IoStatus.Status);
+
+        return STATUS_MORE_PROCESSING_REQUIRED;
+    }
+
+    if (Irp->IoStatus.Information > 0)
+    {
+        if (DeviceExtension->DeviceRelations != NULL && DeviceExtension->DeviceRelations->Count > 0)
+        {
+            // find PDO related to the reportID
+            {
+                CollectionNumber = HidClassPDO_GetReportDescriptionByReportID(&DeviceExtension->Common.DeviceDescription, DeviceExtension->InputBuffer[0])->CollectionNumber;
+
+                for (int i = 0; i < DeviceExtension->DeviceRelations->Count; i++)
+                {
+                    PDODeviceExtension = DeviceExtension->DeviceRelations->Objects[i]->DeviceExtension;
+                    ASSERT(PDODeviceExtension->Common.IsFDO == FALSE);
+
+                    if (PDODeviceExtension->CollectionNumber == CollectionNumber)
+                    {
+                        break;
+                    }
+                    else
+                    {
+                        PDODeviceExtension = NULL;
+                    }
+                }
+            }
+
+            ASSERT(PDODeviceExtension != NULL);
+
+
+            InitializeListHead(&completeIRP);
+
+            KeAcquireSpinLock(&PDODeviceExtension->ReadLock, &CurrentIRQL);
+
+            // check if any IRP waiting data
+            if (IsListEmpty(&PDODeviceExtension->PendingIRPList) == FALSE)
+            {
+                PUCHAR address = NULL;
+                PIRP tempIRP = NULL;
+
+                while (!IsListEmpty(&PDODeviceExtension->PendingIRPList))
```

Comment:
do ... while ?

Reply:
why? more like I just need to move this two variables on the top of this function, and remove "if"

Reply:
The "while (!IsListEmpty(...))" is a better construction.